java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java

```diff
@@ -21,66 +21,65 @@
+  public void visitNode(Tree tree) {
+    SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
+    List<CaseGroupTree> caseGroupTrees = switchStatementTree.cases();
+    CFG cfg = CFG.buildCFG(Collections.singletonList(tree), true);
+    CFG.Block entry = cfg.entry();
+    List<CFG.Block> cfgCases = new ArrayList<>(entry.successors());
+
+    if (!hasDefaultClause(switchStatementTree)) {
+      cfgCases.remove(cfgCases.size() - 1);
+    }
 
-          // Reverse the body as commonly the unconditional exit will be at the end of the body.
-        if (Lists.reverse(caseGroup.body()).stream().noneMatch(SwitchCaseWithoutBreakCheck::isUnconditionalExit) && !intentionalFallThrough(switchStatement, caseGroup)) {
-          context.reportIssue(this, caseLabel, "End this switch case with an unconditional break, return or throw statement.");
-        }
-      });
+    List<CFG.Block> cases = Lists.reverse(cfgCases);
 
-    super.visitSwitchStatement(switchStatement);
+    IntStream.range(1, cases.size())
+      .filter(i -> cases.get(i).predecessors().stream().anyMatch(predecessor -> !tree.equals(predecessor.terminator())))
+      .filter(i -> !intentionalFallThrough(caseGroupTrees.get(i - 1), caseGroupTrees.get(i)))
+      .mapToObj(i -> caseGroupTrees.get(i - 1).labels())
+      .map(caseGroupLabels -> caseGroupLabels.get(caseGroupLabels.size() - 1))
+      .forEach(label -> reportIssue(label, "End this switch case with an unconditional break, return or throw statement."));
   }
 
-  private static boolean intentionalFallThrough(SwitchStatementTree switchStatement, CaseGroupTree caseGroup) {
-    FallThroughCommentVisitor visitor = new FallThroughCommentVisitor();
+  private static boolean hasDefaultClause(SwitchStatementTree switchStatement) {
+    return switchStatement.cases().stream()
+      .flatMap(caseGroupTree -> caseGroupTree.labels().stream())
+      .anyMatch(caseLabelTree -> caseLabelTree.caseOrDefaultKeyword().text().equals("default"));
+  }
+
+  private static boolean intentionalFallThrough(CaseGroupTree caseGroup, CaseGroupTree nextCaseGroup) {
     // Check first token of next case group when comment is last element of case group it is attached to next group.
-    CaseGroupTree nextCaseGroup = switchStatement.cases().get(switchStatement.cases().indexOf(caseGroup) + 1);
+    FallThroughCommentVisitor visitor = new FallThroughCommentVisitor();
     List<Tree> treesToScan = ImmutableList.<Tree>builder().addAll(caseGroup.body()).add(nextCaseGroup.firstToken()).build();
     visitor.scan(treesToScan);
     return visitor.hasComment;
   }
 
   private static class FallThroughCommentVisitor extends SubscriptionVisitor {
-
-    private static final Pattern FALL_THROUGH_PATTERN = Pattern.compile("falls?\\-?thro?u[gh]?", Pattern.CASE_INSENSITIVE);
+    private static final Pattern FALL_THROUGH_PATTERN = Pattern.compile("falls?[\\-\\s]?thro?u[gh]?", Pattern.CASE_INSENSITIVE);
```

Comment:
This change should be documented in a separate JIRA ticket.

Reply:
Created SONARJAVA-2717 for this